### ISSUE TITLE
[FEATURE] HTTP response as second return value of functions of the `opensearchapi` package

### ISSUE THREAD

**Issue by ste93cry** [opensearch-project/opensearch-go#619]
### Is your feature request related to a problem?
In a project I'm working on, I have some integration tests that are testing some code that relies on the functions of the `opensearchapi` package. However, when it's time to assert that the output is equal to the expectations, it's not possible to write a simple `require.Equal(t, wantResp, resp)` because the response object returned by the client includes the `response *opensearch.Response` field, which is hard to predict.

### What solution would you like?
The solution I propose is to return the HTTP response as second return value, instead of inside the API response object. In this way, `SomeResp` becomes somewhat predictable, as it would no longer contain a field whose value is unknown until the request is actually made. In this way, code like the one below is possible without much effort:

```go
resp, err := s.opensearchClient.SomeFunc(context.Background(), SomeReq{
	// Params
})

want := &elastic_types.SomeResp{
	// Expected values
}

require.NoError(t, err)
require.Equal(t, want, resp)
```

Roughly, the changes would be these ones:

```diff
type SomeResp struct {
-	response *opensearch.Response
}

-func (c *Client) SomeFunc(ctx context.Context, req SomeReq) (*SomeResp, error) {
+func (c *Client) SomeFunc(ctx context.Context, req SomeReq) (*SomeResp, *opensearch.Response, error) {
	var (
		data SomeResp
-		err  error
	)

-	if data.response, err = c.do(ctx, req, &data); err != nil {
+	resp, err := c.do(ctx, req, &data)
+	if err != nil {
-		return &data, err
+		return nil, resp, err
+	}

-	return &data, nil
+	return &data, resp, nil
}
```

### What alternatives have you considered?
An alternative solution might be to compare each field of the response object separately. However, this doesn't scale well when there are a lot of fields and makes the code longer and harder to read and maintain.

### Do you have any additional context?
As this is a breaking change, the downside is that it might take quite some time to be released. I would also be open to contribute this change myself, if accepted.

**Comment by Jakob3xD** [comment]
We made the `opensearch.Response` "hidden" behind the `Inspect()` as most users don't need this response and changing this because of some test to look better ... idk.

We are already trying to find missing fields in our tests with json compare. So you might consider using some sort of json compare.
https://github.com/opensearch-project/opensearch-go/blob/bd5d233600802d031ed57b0eca34683f184327ce/internal/test/helper.go#L123

**Comment by ste93cry** [comment]
The problem is not that the field is "hidden" behind an interface, but rather that it is impossible to compare the two structs without resorting to hacks like the one you are mentioning. As you said, a user is often not interested in the raw response, so writing a simple assertion that compares the returned response should work without much hassle. Using an helper like the one you linked works, but it's just a way to solve the problem that the single responsibility principle is broken in this context. I took the idea of returning the response as 2nd value from the Datadog SDK, and I find it quite clear and nice. Returning multiple values is idiomatic in Go, and it makes perfectly clear what is being returned by just looking at the function signature.

**Comment by dblock** [comment]
@ste93cry want to give a PR a shot so we can look at it?

Generally I like explicit returns personally too, but I am concerned about backwards compatibility as well.

**Comment by Jakob3xD** [comment]
> it's not possible to write a simple `require.Equal(t, wantResp, resp)` because the response object returned by the client includes the `response *opensearch.Response` field, which is hard to predict.

Have you tried using `EqualExportedValues()`? Looks like what you want.
https://pkg.go.dev/github.com/stretchr/testify@v1.9.0/require#EqualExportedValues

**Comment by ste93cry** [comment]
I tested your solution and it works, definitely. However, my 2 cents, I still think there is no real reason to couple the HTTP response with an object that represents the deserialized data. If backwards compatibility is the only concern here, I wouldn't really bother too much, as fixing it is quite easy on a practical level.

**Comment by ste93cry** [comment]
>as it is not relevant in most cases
>So why return something you barely need.

I agree on this to a certain degree: I found some cases where the definition of the SDK types are not complete, meaning that not all fields are available. In those cases, having access to the response is definitely useful. Another common use case for the response is debugging. Note I'm not saying that the way to access the response matters here, but they are just two examples of why in some cases the HTTP response is relevant.

This part of a [comment](https://github.com/opensearch-project/opensearch-go/issues/65#issuecomment-1533260393) in that discussion exactly explains why I think that the current solution is suboptimal:

>This is awkward because it has hidden magic

**Comment by dblock** [comment]
I'll try to help here to break a tie ;)

@Jakob3xD, absent of any existing code, if we were writing this library from scratch, do you agree that it's better to return the underlying response explicitly like suggested here? If yes, the pros of the change is that it makes access to the underlying response explicit, and the cons is that it's a breaking change. 

If we can all agree to this, then we need to decide when we take the change.
 
My 0.02c is that we should merge a breaking change like this when we want to make other major breaking changes. As an example, @nhtruong rewrote the opensearch-js client in TypeScript and rolled out a new code generator from API spec in https://github.com/opensearch-project/opensearch-js/issues/803, so together that represented a significant enough improvement to make non-backwards-compatible API changes.

**Comment by nhtruong** [comment]
@dblock small correction: I rewrote the client to consolidate some features but it's still in the common JS. Only the generator is in TS. Your point still stands though. You will most likely introduce breaking changes when API is switched to generated code, so might as well take advantage of the opportunity to fix other glaring issues that also require a major version bump. We had discussion on what to be included in the rewrite [here](https://github.com/opensearch-project/opensearch-js/issues/803) for reference.

**Comment by ste93cry** [comment]
> when we want to make other major breaking changes

A bit off-topic, but if we want to talk about another major breaking change that would level up the SDK is providing users with typed request objects. Right now, when using the `opensearchapi` package, you're left to look at the documentation of each API, understand which fields it needs, write some objects representing the request body and then serialize them to JSON because that's what the SDK expects. To be honest, what's the point of using the SDK then? If we consider the improvements that can be made to this package, most will be breaking changes in a way or another imho.

**Comment by dblock** [comment]
> A bit off-topic, but if we want to talk about another major breaking change that would level up the SDK, we should consider providing users with typed request objects

100%. Someone (you?) should take on https://github.com/opensearch-project/opensearch-go/issues/284

**Comment by sean-** [comment]
This is resolved in two complementary ways - the v5preview codegen (#804) plus the generic `opensearch.Do[T]` helper (#808). Together they give you both the testability fix and the explicit-response shape this issue proposed.

## 1. `opensearch.Do[T]` lands the exact signature you proposed

The generic helper has the shape your diff was asking for:

```go
// opensearch/opensearch.go
func Do[T any](ctx context.Context, c *Client, method string, req Request, dataPointer *T) (*Response, error)
```

The HTTP response is returned explicitly as the second value, the decoded body lands in the caller's `*T`, and the two are no longer entangled inside a struct. In practice:

```go
var resp CatCountResp
httpResp, err := opensearch.Do(ctx, client, http.MethodGet, &CatCountReq{Index: []string{"my-index"}}, &resp)
require.NoError(t, err)
require.Equal(t, want, resp)                          // clean compare, no hidden field
require.Equal(t, http.StatusOK, httpResp.StatusCode)
```

That's essentially the rewrite you sketched, factored as a single generic helper rather than widening every method signature in the API surface. It also has compile-time pointer enforcement (the type system guarantees `*T`) and a `NoBody` marker for endpoints that intentionally skip JSON decoding.

## 2. v5preview generated wrappers hide the response field

For callers who prefer the `(*Resp, error)` convenience path, the v5preview codegen (#804) makes the embedded HTTP response **unexported** with an explicit `Inspect()` accessor:

```go
// v5preview/opensearchapi/cat-count_gen.go
type CatCountResp struct {
    // ... typed fields ...
    response *opensearch.Response   // unexported
}

func (r CatCountResp) Inspect() Inspect {
    return Inspect{Response: r.response}
}
```

This lands two things you cared about in the original report:

- `require.EqualExportedValues(t, want, resp)` (which @Jakob3xD pointed to) now Just Works because `response` is unexported and skipped by that assertion.
- The HTTP response is no longer "hidden magic" inside the struct - it's reachable via an explicit `resp.Inspect()` for debugging or missing-field workarounds.

## How to choose

| Use case | Recommended path |
|---|---|
| Test code asserting decoded body equality | `opensearch.Do[T]` - cleanest compare, no `EqualExportedValues` needed |
| Production code that just wants typed responses | The generated v5preview client - same convenience as before |
| Custom requests not covered by the generated client | `opensearch.Do[T]` - was built specifically for this |

Closing as resolved. If `Do[T]` doesn't fit your code shape for some reason, please open a fresh issue with the specific snippet and we'll take another look.
